src/libutil/logging-diffs.cc

```diff
@@ -0,0 +1,228 @@
+#include "logging.hh"
+#include "logging-diffs.hh"
+#include "util.hh"
+#include "config.hh"
+
+#include <atomic>
+#include <chrono>
+#include <iostream>
+#include <mutex>
+#include <nlohmann/json.hpp>
+#include <thread>
+
+using json = nlohmann::json;
+
+namespace nix {
+
+void addFields(json & json, const Logger::Fields & fields)
+{
+    if (fields.empty()) return;
+    auto & arr = json["fields"] = json::array();
+    for (auto & f : fields)
+        if (f.type == Logger::Field::tInt)
+            arr.push_back(f.i);
+        else if (f.type == Logger::Field::tString)
+            arr.push_back(f.s);
+        else
+            abort();
+}
+
+void to_json(json & j, const NixMessage & m) {
+    j = json{ {"level", m.level} };
+
+    if (m.line.has_value()) j["line"] = m.line.value();
+    if (m.column.has_value()) j["column"] = m.column.value();
+    if (m.file.has_value()) j["file"] = m.file.value();
+
+    if (m.trace.has_value()) j["trace"] = m.trace.value();
+
+    if (!m.msg.empty()) j["msg"] = m.msg;
+    if (!m.raw_msg.empty()) j["raw_msg"] = m.raw_msg;
+}
+
+void to_json(json & j, const ActivityState & as) {
+    j = json{ {"is_complete", as.is_complete}, {"type", as.type}, {"text", as.text} };
+    addFields(j, as.fields);
+}
+
+void to_json(json & j, const NixBuildState & s) {
+    j = json{ {"messages", s.messages} };
+
+    j["activities"] = json(json::value_t::object);
+    for (const auto& [key, value] : s.activities) {
+        j["activities"][std::to_string(key)] = value;
+    }
+}
+
+void add_pos_to_message(NixMessage & msg, std::shared_ptr<AbstractPos> pos)
+{
+    if (pos) {
+        msg.line = pos->line;
+        msg.column = pos->column;
+        std::ostringstream str;
+        pos->print(str);
+        msg.file = str.str();
+    } else {
+        msg.line = std::nullopt;
+        msg.column = std::nullopt;
+        msg.file = std::nullopt;
+    }
+}
+
+void pos_to_json(json & json, std::shared_ptr<AbstractPos> pos)
+{
+    if (pos) {
+        json["line"] = pos->line;
+        json["column"] = pos->column;
+        std::ostringstream str;
+        pos->print(str);
+        json["file"] = str.str();
+    } else {
+        json["line"] = nullptr;
+        json["column"] = nullptr;
+        json["file"] = nullptr;
+    }
+}
+
+struct DiffLogger : Logger {
+    Logger & prevLogger;
+
+    NixBuildState state;
```

Comment:
This should probably be `Sync<NixBuildState> state` and then you can get rid of `std::mutex lock` and all that.

Reply:
Done